docs/spec/overload.rst

```diff
@@ -99,8 +109,376 @@ variable is not sufficient.
+subtype of another), it uses the wider return type. Only if there is no
+consistency relationship between return types does it consider it an
+"ambiguous" situation and turns it into an Any. This produces better
+results for users of language servers, but it doesn't strictly preserve
+the gradual guarantee. I'm willing to abandon this in favor of a
+strict consistency check.]
+
+
+Step 6: Choose the first remaining candidate overload as the winning
+match. Evaluate it as if it were a non-overloaded function call and stop.
+
+
+Example 1::
+
+  @overload
+  def example1(x: int, y: str) -> int: ...
+  @overload
+  def example1(x: str) -> str: ...
+
+  example1()  # Error in step 1: no plausible overloads
+  example1(1, "")  # Step 1 eliminates second overload
+  example1("")  # Step 1 eliminates first overload
+
+  example1("", "")  # Error in step 2: no compatible overloads
+  example1(1)  # Error in step 2: no compatible overloads
+
+
+Example 2::
+
+  @overload
+  def example2(x: int, y: str, z: int) -> str: ...
+  @overload
+  def example2(x: int, y: int, z: int) -> int: ...
+
+  def test(val: str | int):
+      # In this example, argument type expansion is
+      # performed on the first two arguments. Expansion
+      # of the third is unnecessary.
+      r1 = example2(1, val, 1)
+      reveal_type(r1)  # Should reveal str | int
+
+      # Here, the types of all three arguments are expanded
+      # without success.
+      example2(val, val, val)  # Error in step 3
+
+
+Example 3::
+
+  @overload
+  def example3(x: int) -> int: ...
+  @overload
+  def example3(x: int, y: int) -> tuple[int, int]: ...
+  @overload
+  def example3(*args: int) -> tuple[int, ...]: ...
+
+  def test():
+      # Step 1 eliminates second overload. Step 4 and
+      # step 5 do not apply. Step 6 picks the first
+      # overload.
+      r1 = example3(1)
+      reveal_type(r1)  # Should reveal int
+
+      # Step 1 eliminates first overload. Step 4 and
+      # step 5 do not apply. Step 6 picks the second
+      # overload.
+      r2 = example3(1, 2)
+      reveal_type(r2)  # Should reveal tuple[int, int]
+
+      # Step 1 doesn't eliminate any overloads. Step 4
+      # picks the third overload.
+      val = [1, 2, 3, 4]
+      r3 = example3(*val)
+      reveal_type(r3)  # Should reveal tuple[int, ...]
+
+
+Example 4::
+
+  @overload
+  def example4(x: list[int], y: int) -> int: ...
+  @overload
+  def example4(x: list[str], y: str) -> int: ...
+  @overload
+  def example4(x: int, y: int) -> list[int]: ...
+
+  def test(v1: list[Any], v2: Any):
+      # Step 4 eliminates third overload. Step 5
+      # determines that first and second overloads
+      # both apply and are ambiguous due to Any, but
+      # return types are consistent.
+      r1 = example4(v1, 1)
+      reveal_type(r1)  # Should reveal int
+
+      # Step 4 eliminates third overload. Step 5
+      # determines that first and second overloads
```

Comment:
I think the second overload can't apply (and should be eliminated in step 2) because `[1]` is not a `list[str]`.